File: src/utils/loanUtils.test.js
import { describe, it, expect } from 'vitest';
import {
  formatCurrency,
  formatDate,
  formatPercent,
  generateSchedule,
  getScheduleWithStatus,
  getLoanStats,
  assessBorrower,
} from './loanUtils';

describe('loanUtils', () => {
  describe('formatting', () => {
    it('formats currency properly', () => {
      expect(formatCurrency(1250.5)).toBe('$1,250.50');
      expect(formatCurrency(0)).toBe('$0.00');
    });

    it('formats dates properly', () => {
      expect(formatDate('2025-01-15')).toBe('Jan 15, 2025');
      expect(formatDate('')).toBe('N/A');
    });

    it('formats percentage properly', () => {
      expect(formatPercent(12.5)).toBe('12.5%');
    });
  });

  describe('generateSchedule', () => {
    it('generates standard amortizing schedule correctly', () => {
      const schedule = generateSchedule(1000, 12, 12, '2025-01-01');
      expect(schedule).toHaveLength(12);
      expect(schedule[0].due_date).toBe('2025-02-01');
      expect(schedule[0].amount).toBeGreaterThan(0);
      expect(schedule[11].balance).toBe(0);
    });

    it('handles 0% interest rate gracefully', () => {
      const schedule = generateSchedule(1200, 0, 12, '2025-01-01');
      expect(schedule).toHaveLength(12);
      expect(schedule[0].amount).toBe(100);
      expect(schedule[0].interest).toBe(0);
      expect(schedule[11].balance).toBe(0);
    });
  });

  describe('getScheduleWithStatus', () => {
    it('marks installments as paid or partial based on payments', () => {
      const schedule = [
        { installmentNo: 1, due_date: '2025-02-01', amount: 100 },
        { installmentNo: 2, due_date: '2025-03-01', amount: 100 },
      ];
      const payments = [{ amount: 150 }];
      const result = getScheduleWithStatus(schedule, payments, 'active', '2025-01-15');

      expect(result[0].status).toBe('paid');
      expect(result[0].paidAmount).toBe(100);
      expect(result[1].status).toBe('partial');
      expect(result[1].paidAmount).toBe(50);
      expect(result[1].remainingAmount).toBe(50);
    });

    it('marks past-due pending installments as overdue', () => {
      const schedule = [{ installmentNo: 1, due_date: '2025-01-01', amount: 100 }];
      const payments = [];
      const result = getScheduleWithStatus(schedule, payments, 'active', '2025-02-01');

      expect(result[0].status).toBe('overdue');
    });
  });

  describe('getLoanStats', () => {
    it('calculates totals, balance, and progress percentage correctly', () => {
      const loan = {
        principal: 1000,
        status: 'active',
        schedule: [
          { installmentNo: 1, due_date: '2025-02-01', amount: 100 },
          { installmentNo: 2, due_date: '2025-03-01', amount: 100 },
        ],
        payments: [{ amount: 100 }],
      };

      const stats = getLoanStats(loan);
      expect(stats.totalDisbursed).toBe(1000);
      expect(stats.totalPaid).toBe(100);
      expect(stats.outstandingBalance).toBe(100);
      expect(stats.progressPct).toBe(50);
    });
  });

  describe('assessBorrower', () => {
    it('assesses risk rating and DTI properly', () => {
      const borrower = {
        credit_score: 80,
        monthly_income: 4000,
      };
      const loans = [
        {
          status: 'active',
          schedule: [{ amount: 400 }],
        },
        {
          status: 'completed',
        },
      ];

      const assessment = assessBorrower(borrower, loans);
      expect(assessment.dtiPct).toBe(10);
      expect(assessment.riskRating).toBe('low');
      expect(assessment.completedCount).toBe(1);
    });

    it('assigns high risk rating for defaulted loans or high DTI', () => {
      const borrower = {
        credit_score: 50,
        monthly_income: 1000,
      };
      const loans = [{ status: 'defaulted' }];

      const assessment = assessBorrower(borrower, loans);
      expect(assessment.riskRating).toBe('high');
    });
  });
});

File: flutter_app/test/loan_utils_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:microlend/models/borrower.dart';
import 'package:microlend/models/loan.dart';
import 'package:microlend/models/payment.dart';
import 'package:microlend/models/schedule_installment.dart';
import 'package:microlend/utils/loan_utils.dart';

void main() {
  group('LoanUtils Formatting', () {
    test('formatCurrency returns formatted string with selected currency symbol', () {
      expect(LoanUtils.formatCurrency(1250.5, 'USD'), '\$1,250.50');
      expect(LoanUtils.formatCurrency(1250.5, 'EUR'), '€1,250.50');
      expect(LoanUtils.formatCurrency(1250.5, 'PHP'), '₱1,250.50');
      expect(LoanUtils.formatCurrency(1250.5, 'GBP'), '£1,250.50');
      expect(LoanUtils.formatCurrency(0.0), '₱0.00');
    });

    test('formatDate formats date properly', () {
      expect(LoanUtils.formatDate('2025-01-15'), 'Jan 15, 2025');
      expect(LoanUtils.formatDate(''), 'N/A');
    });

    test('formatPercent formats percentage string', () {
      expect(LoanUtils.formatPercent(12.5), '12.5%');
    });
  });

  group('LoanUtils Upfront Deduction', () {
    test('calculates upfront deduction correctly for percent and fixed types', () {
      expect(LoanUtils.calculateUpfrontDeduction(1000.0, 'none', 50.0), 0.0);
      expect(LoanUtils.calculateUpfrontDeduction(1000.0, 'fixed', 50.0), 50.0);
      expect(LoanUtils.calculateUpfrontDeduction(1000.0, 'percent', 5.0), 50.0);
      expect(LoanUtils.calculateUpfrontDeduction(1000.0, 'fixed', 1500.0), 1000.0);
    });

    test('calculates net disbursed correctly', () {
      expect(LoanUtils.calculateNetDisbursed(1000.0, 'none', 0.0), 1000.0);
      expect(LoanUtils.calculateNetDisbursed(1000.0, 'fixed', 50.0), 950.0);
      expect(LoanUtils.calculateNetDisbursed(1000.0, 'percent', 2.5), 975.0);
    });

    test('getLoanStats uses net disbursed for totalDisbursed', () {
      final loan = Loan(
        id: 'l1',
        borrowerId: 'b1',
        principal: 1000.0,
        interestRate: 12.0,
        termMonths: 12,
        purpose: 'Test',
        status: 'active',
        disbursementDate: '2025-01-01',
        upfrontDeductionType: 'percent',
        upfrontDeductionValue: 5.0,
        schedule: [],
        payments: [],
        notes: '',
      );

      final stats = LoanUtils.getLoanStats(loan);
      expect(stats.totalDisbursed, 950.0);
    });
  });

  group('LoanUtils.generateSchedule', () {
    test('generates 12 month amortizing loan schedule correctly', () {
      final schedule = LoanUtils.generateSchedule(1000.0, 12.0, 12, '2025-01-01');
      expect(schedule.length, 12);
      expect(schedule[0].dueDate, '2025-02-01');
      expect(schedule[0].amount, greaterThan(0.0));
      expect(schedule[11].balance, 0.0);
    });

    test('handles 0% interest rate gracefully', () {
      final schedule = LoanUtils.generateSchedule(1200.0, 0.0, 12, '2025-01-01');
      expect(schedule.length, 12);
      expect(schedule[0].amount, 100.0);
      expect(schedule[0].interest, 0.0);
      expect(schedule[11].balance, 0.0);
    });

    test('LoanUtils.generateSchedule handles flat "5-6" interest method correctly', () {
      // 1000 principal, 20% flat rate (add-on), 10 weekly installments
      final schedule = LoanUtils.generateSchedule(
        1000.0,
        20.0,
        10,
        '2026-01-01',
        repaymentFrequency: 'weekly',
        interestMethod: 'flat',
      );

      expect(schedule.length, 10);
      // Total interest = 1000 * 20% = 200. Total repayable = 1200. Per weekly installment = 120.
      expect(schedule[0].amount, 120.0);
      expect(schedule[0].principal, 100.0);
      expect(schedule[0].interest, 20.0);
      expect(schedule[0].dueDate, '2026-01-08');
      expect(schedule[9].dueDate, '2026-03-12');
    });

    test('LoanUtils.generateSchedule handles interest-only balloon interest method', () {
      final schedule = LoanUtils.generateSchedule(
        1200.0,
        12.0,
        12,
        '2026-01-01',
        repaymentFrequency: 'monthly',
        interestMethod: 'interest_only',
      );

      expect(schedule.length, 12);
      // Monthly interest = 1200 * (12%/12) = 12.0
      expect(schedule[0].amount, 12.0);
      expect(schedule[0].principal, 0.0);
      expect(schedule[0].interest, 12.0);

      // Final installment includes full principal balloon (1200 + 12 = 1212)
      expect(schedule[11].amount, 1212.0);
      expect(schedule[11].principal, 1200.0);
      expect(schedule[11].interest, 12.0);
    });

    test('LoanUtils.generateSchedule handles one-time lump sum payment', () {
      final schedule = LoanUtils.generateSchedule(
        500.0,
        10.0,
        1,
        '2026-01-01',
        repaymentFrequency: 'monthly',
        interestMethod: 'one_time',
      );

      expect(schedule.length, 1);
      expect(schedule[0].amount, 550.0);
      expect(schedule[0].principal, 500.0);
      expect(schedule[0].interest, 50.0);
      expect(schedule[0].dueDate, '2026-02-01');
    });
  });

  group('LoanUtils.getScheduleWithStatus', () {
    test('marks installments as paid or partial based on payments', () {
      final schedule = [
        ScheduleInstallment(installmentNo: 1, dueDate: '2025-02-01', amount: 100.0, principal: 90.0, interest: 10.0, balance: 900.0),
        ScheduleInstallment(installmentNo: 2, dueDate: '2025-03-01', amount: 100.0, principal: 91.0, interest: 9.0, balance: 809.0),
      ];
      final payments = [Payment(id: 'p1', date: '2025-01-10', amount: 150.0, method: 'Cash', note: '')];
      final result = LoanUtils.getScheduleWithStatus(schedule, payments, 'active', DateTime.parse('2025-01-15'));

      expect(result[0].status, 'paid');
      expect(result[0].paidAmount, 100.0);
      expect(result[1].status, 'partial');
      expect(result[1].paidAmount, 50.0);
      expect(result[1].remainingAmount, 50.0);
    });
  });

  group('LoanUtils.assessBorrower', () {
    test('calculates credit score, DTI and risk rating', () {
      final borrower = Borrower(
        id: 'b1',
        fullName: 'Elena',
        email: '',
        phone: '',
        address: '',
        idNumber: '',
        employment: '',
        monthlyIncome: 4000.0,
        creditScore: 80,
        riskRating: 'low',
        notes: '',
      );
      final loans = [
        Loan(
          id: 'l1',
          borrowerId: 'b1',
          principal: 1000.0,
          interestRate: 10.0,
          termMonths: 12,
          purpose: 'Test',
          status: 'active',
          disbursementDate: '2025-01-01',
          schedule: [ScheduleInstallment(installmentNo: 1, dueDate: '2025-02-01', amount: 400.0, principal: 390.0, interest: 10.0, balance: 600.0)],
          payments: [],
          notes: '',
        ),
      ];

      final assessment = LoanUtils.assessBorrower(borrower, loans);
      expect(assessment.dtiPct, 10);
      expect(assessment.riskRating, 'low');
    });

    test('DTI normalization converts daily repayment frequency to monthly debt (~30.4x)', () {
      final borrower = Borrower(
        id: 'b2',
        fullName: 'Daily Borrower',
        email: '',
        phone: '',
        address: '',
        idNumber: '',
        employment: '',
        monthlyIncome: 3000.0,
        creditScore: 75,
        riskRating: 'low',
        notes: '',
      );

      final dailyLoan = Loan(
        id: 'daily_l1',
        borrowerId: 'b2',
        principal: 1000.0,
        interestRate: 12.0,
        termMonths: 0,
        repaymentFrequency: 'daily',
        termCount: 30,
        purpose: 'Daily Test',
        status: 'active',
        disbursementDate: '2026-01-01',
        schedule: [ScheduleInstallment(installmentNo: 1, dueDate: '2026-01-02', amount: 35.0, principal: 33.0, interest: 2.0, balance: 967.0)],
        payments: [],
        notes: '',
      );

      final assessment = LoanUtils.assessBorrower(borrower, [dailyLoan]);
      // Daily 35.0 * 30.4167 = ~1064.58 monthly debt. DTI = 1064.58 / 3000 = ~35%
      expect(assessment.monthlyDebt, closeTo(1064.58, 1.0));
      expect(assessment.dtiPct, 35);
    });
  });

  group('LoanUtils.computeEarlyPayoffAmount and validateLoanParams', () {
    test('early payoff amount for reducing loan is less than sum of remaining scheduled installments', () {
      final schedule = LoanUtils.generateSchedule(10000.0, 12.0, 12, '2026-01-01');
      final loan = Loan(
        id: 'reducing_early_payoff',
        borrowerId: 'b1',
        principal: 10000.0,
        interestRate: 12.0,
        termMonths: 12,
        repaymentFrequency: 'monthly',
        interestMethod: 'reducing',
        termCount: 12,
        purpose: 'Early Payoff Test',
        status: 'active',
        disbursementDate: '2026-01-01',
        schedule: schedule,
        payments: [],
        notes: '',
      );

      final stats = LoanUtils.getLoanStats(loan, DateTime.parse('2026-01-15'));
      // Total scheduled includes 12 months of interest (~10,661.85). Early payoff at day 14 should be ~10,040 (principal + 14 days interest)
      expect(stats.payoffAmount, lessThan(stats.totalScheduled));
      expect(stats.payoffAmount, greaterThan(10000.0));
    });

    test('validateLoanParams rejects negative interest, negative penalty, and excessive tenure', () {
      expect(LoanUtils.validateLoanParams(principal: -100, interestRate: 10, termCount: 6, repaymentFrequency: 'monthly', penaltyValue: 0), isNotNull);
      expect(LoanUtils.validateLoanParams(principal: 1000, interestRate: -5, termCount: 6, repaymentFrequency: 'monthly', penaltyValue: 0), isNotNull);
      expect(LoanUtils.validateLoanParams(principal: 1000, interestRate: 10, termCount: 6, repaymentFrequency: 'monthly', penaltyValue: -10), isNotNull);
      expect(LoanUtils.validateLoanParams(principal: 1000, interestRate: 10, termCount: 100, repaymentFrequency: 'monthly', penaltyValue: 0), isNotNull);
      expect(LoanUtils.validateLoanParams(principal: 1000, interestRate: 10, termCount: 12, repaymentFrequency: 'monthly', penaltyValue: 0), isNull);
    });
  });
}
